### PR TITLE
Better environment shutdown

### PR DESCRIPTION
**Pull request by chriselion**
While working on the timer code in https://github.com/Unity-Technologies/ml-agents/pull/2198 I noticed we weren't executing code at shutdown (namely so that I could write out the timer data).

This fixes two issues:
1) Don't kill the subprocess right away, give it a chance to shutdown. This should happen in the Academy code when the QUIT command is received*:
https://github.com/Unity-Technologies/ml-agents/blob/c916801475950ef92016b51c3fe5806ac60afc90/UnitySDK/Assets/ML-Agents/Scripts/Academy.cs#L536-L544
2) Launch the executable process with `start_new_session=True`. Without this, hitting Ctrl-C would propagate the signal (SIGINT) to the subprocess, so it dies before getting a chance to do anything.

Related to 2), also nicely format the return code (would have saved me some time if I know negative return corresponded to the signal ID).

I'm not sure if we should add configuration parameters for
* shutdown timeout (I'm currently using the communicator's `timeout_wait` which defaults to 30s, but that seems too long)
* whether or not to set start_new_session - I think this should default to True but might need an option to override.

*(NOTE) We don't actually send QUIT here, but when the communicator gets the `400` status that python sends
https://github.com/Unity-Technologies/ml-agents/blob/c916801475950ef92016b51c3fe5806ac60afc90/UnitySDK/Assets/ML-Agents/Scripts/RpcCommunicator.cs#L107-L116
it returns null, so the batcher treats this as a QUIT
https://github.com/Unity-Technologies/ml-agents/blob/c916801475950ef92016b51c3fe5806ac60afc90/UnitySDK/Assets/ML-Agents/Scripts/Batcher.cs#L241-L245

### CODE REVIEW

File: ml-agents-envs/mlagents/envs/subprocess_env_manager.py [hunk 1]
```diff
@@ -119,18 +119,18 @@ def _send_response(cmd_name, payload):
             elif cmd.name == "close":
                 break
     except (KeyboardInterrupt, UnityCommunicationException):
-        print("UnityEnvironment worker: environment stopping.")
+        logger.info(f"UnityEnvironment worker {worker_id}: environment stopping.")
```
Comment (chriselion) [inline ml-agents-envs/mlagents/envs/subprocess_env_manager.py]: Cleaned this up to use `logger` and made it and the debug's below more consistent.

File: ml-agents-envs/mlagents/envs/environment.py [hunk 2]
```diff
@@ -583,7 +592,16 @@ def _close(self):
         self._loaded = False
         self.communicator.close()
         if self.proc1 is not None:
-            self.proc1.kill()
+            # Wait a bit for the process to shutdown, but kill it if it takes too long
+            try:
+                self.proc1.wait(timeout=self.timeout_wait)
```
Comment (surfnerd) [inline ml-agents-envs/mlagents/envs/environment.py]: I think this addresses the main concern I had, which was a user would have to manually close 15 unity processes.  But it looks like you are waiting here to see if they play nice, then drop the hammer if not.

Comment (chriselion) [reply ml-agents-envs/mlagents/envs/environment.py]: True, it'll still eventually get killed. But we're relying on the Academy interpreting the commands and closing the application; a game developer might expect the signals to get propagated.

Maybe we should make Academy listen for signals instead?

**Comment by chriselion** [comment]
@harperj Haven't tried on Windows. The start_new_session parameter only applies to POSIX systems. From https://docs.python.org/3/library/subprocess.html#popen-constructor
> If start_new_session is true the setsid() system call will be made in the child process prior to the execution of the subprocess. (POSIX only)

So it at least shouldn't make anything worse on Windows (I'm not sure if `KeyboardInterrupt` will propagate there anyway)

I think the "wait, else kill" approach is generally better.

File: ml-agents-envs/mlagents/envs/environment.py (hunk 2 above)
Comment (surfnerd) [reply ml-agents-envs/mlagents/envs/environment.py]: ehhh, I don't like the sound of that.

**Comment by harperj** [comment]
I'd recommend at least giving it a test on a Windows machine before merging.  Otherwise 🚢 .

**Comment by chriselion** [comment]
Tested on windows - CTRL-C still gets sent to the executable, but it's no worse (and not sure we can do better)

File: ml-agents-envs/mlagents/envs/environment.py (hunk 2 above)
Comment (chriselion) [reply ml-agents-envs/mlagents/envs/environment.py]: OK, punting on this for now. If it comes up, we'll add a flag to toggle `start_new_session`

**Pull request merged**